src/shared/index.ts

```diff
@@ -0,0 +1,216 @@
+
+			result._rule = run.tool.driver.rules?.[result.ruleIndex ?? -1] // If result.ruleIndex is undefined, that's okay.
+			const message = result._rule?.messageStrings?.[result.message.id ?? -1] ?? result.message
+			result._message = format(message.text || result.message?.text, result.message.arguments) ?? '—'
+			result._markdown = format(message.markdown || result.message?.markdown, result.message.arguments) // No '—', leave undefined if empty.
+
+			result.level = result.level ?? 'warning'
+			result.baselineState = result.baselineState ?? 'new'
+			result._suppression = !result.suppressions || result.suppressions.every(sup => sup.status === 'rejected')
+				? 'not suppressed'
+				: 'suppressed'
+		})
+
+		run._implicitBase = implicitBase?.join('/')
+		run.results?.forEach(result => {
+			result._relativeUri = result._uri?.replace(run._implicitBase ?? '' , '') ?? '' // For grouping, Empty works more predictably than undefined
+		})
+	})
+	log._distinct = mapDistinct(fileAndUris)
+	log._jsonMap = undefined // Free-up memory.
+}
+
+/*
+TfLoc
+   location: Loc
+
+Result
+   locations: Loc[]
+
+Loc
+   Message
+   PhyLoc
+      ArtLoc: Uri, Index
+      Region
+
+Run.artifacts: Art[]
+   location: ArtLoc
+   contents: ArtCon
+*/
+export function parseLocation(result: Result, loc?: Location) {
+	const message = loc?.message?.text
+	const [uri, uriContent] = parseArtifactLocation(result, loc?.physicalLocation?.artifactLocation)
+	const region = loc?.physicalLocation?.region
+	return { message, uri, uriContent, region }
+}
+
+export function parseRegion(region: Region | undefined): _Region | undefined {
+	if (!region) return undefined
+
+	const {byteOffset, byteLength} = region
+	if (byteOffset !== undefined && byteLength !== undefined) return [byteOffset, byteLength] as _RegionBytes
+
+	let {startLine, startColumn, endLine, endColumn} = region
+	if (!startLine) return undefined // Lines are 1-based so no need to check undef.
+
+	startLine--
+	if (!startColumn) return startLine
+
+	startColumn--
+	if (endColumn) endColumn--
+	if (endLine) endLine--
+	return [
+		startLine,
+		startColumn,
+		endLine ?? startLine,
+		endColumn ?? (startColumn + 1)
+	] as _RegionStartEndLineCol
+}
+
+// Improve: `result` purely used for `_run.artifacts`.
+export function parseArtifactLocation(result: Result, anyArtLoc: ArtifactLocation | undefined) {
+	if (!anyArtLoc) return [undefined, undefined]
+	const runArt = result._run.artifacts?.[anyArtLoc.index ?? -1]
+	const runArtLoc = runArt?.location
+	const runArtCon = runArt?.contents
+
+	const uri = anyArtLoc.uri ?? runArtLoc?.uri // If index (§3.4.5) is absent, uri SHALL be present.
+	const uriContents = runArtCon?.text || runArtCon?.binary
+			? encodeURI(`sarif:${encodeURIComponent(result._log._uri)}/${result._run._index}/${anyArtLoc.index}/${uri?.file ?? 'Untitled'}`)
+			: undefined
+	return [uri, uriContents]
+}
+
+export const filtersRow = {
+	Level: {
+		'Error': true,
```

Comment:
>true [](start = 11, length = 4)

What do the true\false mean here? (and below)?

Can we create a type to help me understand this?

Reply:
In our repo code, we actually shy away from using bools at all and use a type that is a union of string literals for code readibility.

---
In reply to: [437693237](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r437693237) [](ancestors = 437693237)

Reply:
Show/hide, visible/non-visible. I can change. Any preference on the terms used?  There are downstream effects so this may take a minute.

---
In reply to: [437696855](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r437696855) [](ancestors = 437696855,437693237)

Reply:
Changed.

---
In reply to: [437704442](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r437704442) [](ancestors = 437704442,437696855,437693237)